base/locks-mt.jl

```diff
@@ -61,10 +61,12 @@ Base.assert_havelock(l::SpinLock) = islocked(l) ? nothing : Base.concurrency_vio
 function lock(l::SpinLock)
     while true
         if _get(l) == 0
+            GC.enable_finalizers(false)
```

Comment:
Why can't this be done just before returning?

Reply:
The entire point of this bug fix is that it must be done before lock acquisition

Reply:
For comparison, `jl_mutex_lock` only disables finalizers after `jl_mutex_wait` returns. I thought that would correspond to disabling them just after `if p == 0` here.

Reply:
jl_mutex_lock can get away with doing potentially suspect things because it can ensure they are indistinguishable to the user